src/submodule.c

```diff
@@ -149,6 +149,41 @@ static int find_by_path(const git_config_entry *entry, void *payload)
 	return 0;
 }
 
+static int can_add_submodule(git_repository *repo, const char *path) {
+	int error;
+	git_index *index;
+	git_buf dir = GIT_BUF_INIT;
+
+	if ((error = git_buf_sets(&dir, path)) < 0)
+		return error;
+
+	if ((error = git_path_to_dir(&dir)) < 0)
```

Comment:
You start to leak memory as soon as one of those functions returns an error. Instead of "return error;", we usually have e.g. a "goto error;" where all memory is free'd. E.g.:

```
static int can_add_submodule(git_repository *repo, const char *path)
{
...

if ((error = git_path_to_dir(&dir)) < 0)
    goto out;

...

out:
    git_buf_free(&dir);
    return error;
}
```